File: apps/web/src/shared/api/transport.mjs
// Shared by the React console and the embedded console. WebCrypto never falls
// back to plaintext, and key/challenge material is never persisted or cached.
const encoder = new TextEncoder();
const prefix = "access-gateway/browser-transport/v1";

export class TransportError extends Error {
  constructor(message) { super(message); this.name = "TransportError"; }
}

export function requiresEncryption(method, path) {
  if (method === "GET" && /^\/api\/v1\/sessions\/[^/]+\/terminal$/.test(path)) return true;
  if (method === "POST" && /^\/api\/v1\/sessions\/[^/]+\/terminal\/demo-defaults$/.test(path)) return true;
	if (method === "POST" && (["/api/v1/auth/invitation/preview", "/api/v1/auth/invitation/accept", "/api/v1/auth/mfa", "/api/v1/auth/mfa/enroll", "/api/v1/auth/mfa/confirm", "/api/v1/auth/mfa/verify", "/api/v1/auth/account/mfa/enroll", "/api/v1/auth/account/mfa/recovery-codes", "/api/v1/auth/account/mfa/unbind", "/api/v1/admin/invitations"].includes(path) || /^\/api\/v1\/admin\/invitations\/[^/]+\/resend$/.test(path))) return true;
  if (method === "POST") {
    return ["/api/v1/auth/local/login", "/api/v1/auth/ldap/login", "/api/v1/auth/password", "/api/v1/admin/cloud-accounts", "/api/v1/admin/assets", "/api/v1/admin/users", "/api/v1/admin/settings/audit/certificates", "/api/v1/admin/assets/audit/certificates"].includes(path) ||
      /^\/api\/v1\/admin\/gateways\/[^/]+\/release$/.test(path);
  }
  return method === "PATCH" && (path === "/api/v1/admin/settings" || /^\/api\/v1\/admin\/(cloud-accounts|assets)\/[^/]+$/.test(path));
}

// The WebSocket upgrade has no request body. Seal its first message using the
// same one-time, actor/path-bound contract as sensitive HTTP requests.
export async function sealTerminalStart(path, body, signal) {
  if (!/^\/api\/v1\/sessions\/[a-zA-Z0-9-]+\/terminal$/.test(path)) throw new TransportError("终端请求地址无效。");
  if (!globalThis.crypto?.subtle) throw new TransportError("敏感信息需要加密传输，请使用 HTTPS 或 localhost 访问。");
  const response = await fetch("/api/v1/auth/transport/challenges", {
    method: "POST", credentials: "same-origin", cache: "no-store", redirect: "error", signal,
    headers: { Accept: "application/json", "Content-Type": "application/json" },
    body: JSON.stringify({ method: "GET", path }),
  });
  if (!response.ok) throw new TransportError("无法获取终端加密凭据，请检查登录状态后重试。");
  const challenge = await response.json();
  if (challenge.method !== "GET" || challenge.path !== path || !challenge.id || !challenge.kid || typeof challenge.subject !== "string") {
    throw new TransportError("加密挑战校验失败，请刷新页面后重试。");
  }
  return (await sealPayload(challenge, body)).body;
}

function base64(bytes) {
  let text = "";
  for (let offset = 0; offset < bytes.length; offset += 8192) text += String.fromCharCode(...bytes.subarray(offset, offset + 8192));
  return btoa(text);
}

function unbase64(text) {
  return Uint8Array.from(atob(text), character => character.charCodeAt(0));
}

function aad(challenge, direction) {
  return encoder.encode([prefix, direction, challenge.id, challenge.kid, challenge.method, challenge.path, challenge.subject].join("\n"));
}

async function derive(seedKey, direction) {
  return crypto.subtle.deriveKey({ name: "HKDF", hash: "SHA-256", salt: new Uint8Array(), info: encoder.encode(`${prefix}/${direction}`) },
    seedKey, { name: "AES-GCM", length: 256 }, false, direction === "request" ? ["encrypt"] : ["decrypt"]);
}

export async function sealPayload(challenge, body) {
  if (!globalThis.crypto?.subtle) throw new TransportError("敏感信息需要加密传输，请使用 HTTPS 或 localhost 访问。");
  if (challenge.algorithm !== "RSA-OAEP-256+A256GCM") throw new TransportError("服务端加密协议不兼容，请刷新页面后重试。");
  const seed = crypto.getRandomValues(new Uint8Array(32));
  const plaintext = encoder.encode(JSON.stringify(body));
  try {
    if (plaintext.length > 512 * 1024) throw new TransportError("提交内容过大，请减少内容后重试。");
    const publicDER = unbase64(challenge.public_key.replace(/-----[\w ]+-----|\s/g, ""));
    const publicKey = await crypto.subtle.importKey("spki", publicDER, { name: "RSA-OAEP", hash: "SHA-256" }, false, ["encrypt"]);
    if (publicKey.algorithm.modulusLength < 2048) throw new TransportError("服务端加密密钥不符合要求。");
    const seedKey = await crypto.subtle.importKey("raw", seed, "HKDF", false, ["deriveKey"]);
    const requestKey = await derive(seedKey, "request");
    const responseKey = await derive(seedKey, "response");
    const nonce = crypto.getRandomValues(new Uint8Array(12));
    const ciphertext = await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce, additionalData: aad(challenge, "request"), tagLength: 128 }, requestKey, plaintext);
    const encryptedKey = await crypto.subtle.encrypt({ name: "RSA-OAEP" }, publicKey, seed);
    return {
      body: { challenge_id: challenge.id, envelope: { version: 1, kid: challenge.kid, encrypted_key: base64(new Uint8Array(encryptedKey)), nonce: base64(nonce), ciphertext: base64(new Uint8Array(ciphertext)) } },
      async open(response, status) {
        if (response?.version !== 1 || response.challenge_id !== challenge.id) throw new TransportError("加密响应校验失败，请重试。");
        let decrypted;
        try {
          decrypted = new Uint8Array(await crypto.subtle.decrypt({ name: "AES-GCM", iv: unbase64(response.nonce), additionalData: aad(challenge, `response:${status}`), tagLength: 128 }, responseKey, unbase64(response.ciphertext)));
          return new TextDecoder("utf-8", { fatal: true }).decode(decrypted);
        } catch {
          throw new TransportError("加密响应校验失败，请重试。");
        } finally { decrypted?.fill(0); }
      },
    };
  } catch (error) {
    if (error instanceof TransportError) throw error;
    throw new TransportError("无法建立加密传输，请刷新页面后重试。");
  } finally { seed.fill(0); plaintext.fill(0); }
}

export async function encryptedFetch(url, options = {}) {
  if (!url.startsWith("/api/v1/") || /[#\\\r\n]/.test(url)) throw new TransportError("API 请求地址无效。");
  const method = (options.method || "GET").toUpperCase();
  const path = url.split("?")[0];
  if (!requiresEncryption(method, path)) return fetch(url, options);
  if (!globalThis.crypto?.subtle) throw new TransportError("敏感信息需要加密传输，请使用 HTTPS 或 localhost 访问。");
  // Query strings are excluded from the contract; protected operations carry all
  // arguments in the authenticated JSON body or resource path.
  if (url !== path) throw new TransportError("敏感接口不支持 URL 查询参数。");
  const headers = new Headers(options.headers);
  const challengeHeaders = new Headers({ Accept: "application/json", "Content-Type": "application/json" });
  const challengeResponse = await fetch("/api/v1/auth/transport/challenges", {
    method: "POST", credentials: "same-origin", cache: "no-store", redirect: "error", signal: options.signal,
    headers: challengeHeaders, body: JSON.stringify({ method, path }),
  });
  if (!challengeResponse.ok) return challengeResponse;
  const challenge = await challengeResponse.json();
  if (challenge.method !== method || challenge.path !== path || !challenge.id || !challenge.kid || typeof challenge.subject !== "string") {
    throw new TransportError("加密挑战校验失败，请刷新页面后重试。");
  }
  const sealed = await sealPayload(challenge, options.body === undefined ? {} : JSON.parse(options.body));
  headers.set("Content-Type", "application/json");
  const response = await fetch(url, { ...options, method, headers, credentials: "same-origin", cache: "no-store", redirect: "error", body: JSON.stringify(sealed.body) });
  if (response.headers.get("X-AG-Encrypted") !== "1") {
    if (!response.ok) return response; // CSRF, authorization and envelope errors contain no secrets.
    throw new TransportError("服务端未返回加密响应，请刷新页面或联系管理员。");
  }
  const plaintext = await sealed.open(await response.json(), response.status);
  const responseHeaders = new Headers(response.headers);
  responseHeaders.delete("Content-Length");
  responseHeaders.delete("Content-Encoding");
  return new Response(plaintext, { status: response.status, statusText: response.statusText, headers: responseHeaders });
}

File: internal/httpapi/transport.go
package httpapi

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"net"
	"net/http"
	"regexp"
	"strings"

	"github.com/emicklei/go-restful/v3"
	"github.com/srex-run/access-gateway/internal/securetransport"
	"github.com/srex-run/access-gateway/internal/security"
	"github.com/srex-run/access-gateway/internal/service"
)

type BrowserTransport interface {
	Create(context.Context, securetransport.Binding) (securetransport.Challenge, error)
	Open(context.Context, securetransport.Binding, securetransport.Request) ([]byte, *securetransport.Reply, error)
}

var encryptedResourcePath = regexp.MustCompile(`^/api/v1/(admin/(cloud-accounts|assets)/[^/]+|admin/gateways/[^/]+/release)$`)
var encryptedTerminalPath = regexp.MustCompile(`^/api/v1/sessions/[^/]+/terminal$`)
var encryptedTerminalDemoDefaultsPath = regexp.MustCompile(`^/api/v1/sessions/[^/]+/terminal/demo-defaults$`)

func anonymousIdentityRoute(path string) bool {
	return path == "/api/v1/auth/invitation/preview" || path == "/api/v1/auth/invitation/accept"
}

func pendingMFARoute(path string) bool {
	return path == "/api/v1/auth/mfa" || path == "/api/v1/auth/mfa/enroll" || path == "/api/v1/auth/mfa/confirm" || path == "/api/v1/auth/mfa/verify"
}

var invitationResendPath = regexp.MustCompile(`^/api/v1/admin/invitations/[^/]+/resend$`)

func identityEncryptedRoute(method, path string) bool {
	return method == http.MethodPost && (anonymousIdentityRoute(path) || pendingMFARoute(path) ||
		path == "/api/v1/auth/account/mfa/enroll" || path == "/api/v1/auth/account/mfa/recovery-codes" || path == "/api/v1/auth/account/mfa/unbind" ||
		path == "/api/v1/admin/invitations" || invitationResendPath.MatchString(path))
}

// This is a mandatory wire contract, independent of provider settings and body
// contents. Even edits that preserve existing credentials use the same contract.
func encryptedRoute(method, path string) bool {
	if method == http.MethodGet && encryptedTerminalPath.MatchString(path) {
		return true
	}
	if identityEncryptedRoute(method, path) {
		return true
	}
	if method == http.MethodPost {
		if encryptedTerminalDemoDefaultsPath.MatchString(path) {
			return true
		}
		switch path {
		case "/api/v1/auth/local/login", "/api/v1/auth/ldap/login", "/api/v1/auth/password", "/api/v1/admin/cloud-accounts", "/api/v1/admin/assets", "/api/v1/admin/users", "/api/v1/admin/settings/audit/certificates", "/api/v1/admin/assets/audit/certificates":
			return true
		}
		return encryptedResourcePath.MatchString(path) && !strings.HasPrefix(path, "/api/v1/admin/cloud-accounts/") && !strings.HasPrefix(path, "/api/v1/admin/assets/")
	}
	return method == http.MethodPatch && (path == "/api/v1/admin/settings" ||
		((strings.HasPrefix(path, "/api/v1/admin/cloud-accounts/") || strings.HasPrefix(path, "/api/v1/admin/assets/")) && encryptedResourcePath.MatchString(path)))
}

func (s *Server) transportBinding(request *restful.Request, response *restful.Response, method, path string) (securetransport.Binding, bool) {
	binding := securetransport.Binding{Method: method, Path: path}
	if pendingMFARoute(path) {
		token, ok := s.pendingToken(request, response)
		if !ok {
			return binding, false
		}
		binding.Subject = "mfa:" + security.HashOpaqueToken(token)
		return binding, true
	}
	if path == "/api/v1/auth/local/login" || path == "/api/v1/auth/ldap/login" || anonymousIdentityRoute(path) {
		// There is no authenticated actor yet. Avoid proxy-IP binding: a browser
		// may reach different console replicas between challenge and submission.
		binding.Subject = "anonymous"
		return binding, true
	}
	userID, ok := s.currentUser(request, response)
	if !ok {
		return binding, false
	}
	if permission, protected := routePermission(method, path); protected {
		if err := s.Service.Authorize(request.Request.Context(), userID, permission); err != nil {
			s.writeError(response, err)
			return binding, false
		}
	}
	binding.Subject = "user:" + userID
	if cookie, err := request.Request.Cookie(s.SessionCookieName); err == nil {
		binding.Subject += ":" + security.HashOpaqueToken(cookie.Value)
	}
	return binding, true
}

func transportAddress(request *http.Request) string {
	ip, _, err := net.SplitHostPort(request.RemoteAddr)
	if err != nil {
		return request.RemoteAddr
	}
	return ip
}

func (s *Server) allowTransport(request *restful.Request, response *restful.Response, purpose string) bool {
	allowed, err := s.LoginAttempts.Allow(request.Request.Context(), s.DB,
		security.HashOpaqueToken("browser-transport:"+purpose+":"+transportAddress(request.Request)), 300)
	if err != nil {
		s.writeError(response, err)
		return false
	}
	if !allowed {
		response.Header().Set("Retry-After", "300")
		_ = response.WriteHeaderAndEntity(http.StatusTooManyRequests, map[string]string{"error": "too many encrypted requests"})
	}
	return allowed
}

func (s *Server) createTransportChallenge(request *restful.Request, response *restful.Response) {
	var input struct {
		Method string `json:"method"`
		Path   string `json:"path"`
	}
	if decodeJSONBody(request.Request, &input, false) != nil || len(input.Path) > 512 ||
		strings.ContainsAny(input.Path, "?%#\\\r\n") || strings.Contains(input.Path, "..") || !encryptedRoute(input.Method, input.Path) {
		s.writeError(response, service.ErrValidation)
		return
	}
	binding, ok := s.transportBinding(request, response, input.Method, input.Path)
	if !ok || !s.allowTransport(request, response, "challenge") {
		return
	}
	if s.Transport == nil {
		s.writeError(response, service.ErrNotConfigured)
		return
	}
	challenge, err := s.Transport.Create(request.Request.Context(), binding)
	if err != nil {
		s.writeError(response, err)
		return
	}
	_ = response.WriteEntity(challenge)
}

// Decrypt only after CSRF and authorization checks. The original HTTP body stays
// encrypted for outer middleware. Handler responses (including errors) are
// encrypted before reaching the proxy or response logging middleware.
func (s *Server) transportFilter(request *restful.Request, response *restful.Response, chain *restful.FilterChain) {
	// The terminal handler authenticates/decrypts the first WebSocket message;
	// an HTTP upgrade cannot carry the ordinary encrypted request body.
	if request.Request.Method == http.MethodGet && encryptedTerminalPath.MatchString(request.SelectedRoutePath()) {
		chain.ProcessFilter(request, response)
		return
	}
	if !encryptedRoute(request.Request.Method, request.SelectedRoutePath()) {
		chain.ProcessFilter(request, response)
		return
	}
	if request.Request.URL.RawQuery != "" || request.Request.URL.RawPath != "" ||
		strings.HasSuffix(request.Request.URL.Path, "/") || strings.Contains(request.Request.URL.Path, "//") {
		s.writeError(response, service.ErrValidation)
		return
	}
	binding, ok := s.transportBinding(request, response, request.Request.Method, request.Request.URL.Path)
	if !ok {
		return
	}
	var envelope securetransport.Request
	if decodeJSONBody(request.Request, &envelope, false) != nil || envelope.ChallengeID == "" || envelope.Envelope.Version != 1 {
		_ = response.WriteHeaderAndEntity(http.StatusBadRequest, map[string]string{"error": "encrypted request required; refresh the page and retry"})
		return
	}
	if !s.allowTransport(request, response, "open") {
		return
	}
	if s.Transport == nil {
		s.writeError(response, service.ErrNotConfigured)
		return
	}
	plaintext, reply, err := s.Transport.Open(request.Request.Context(), binding, envelope)
	if err != nil {
		if errors.Is(err, securetransport.ErrInvalid) {
			_ = response.WriteHeaderAndEntity(http.StatusBadRequest, map[string]string{"error": "encrypted request is invalid or expired; retry"})
		} else {
			s.writeError(response, err)
		}
		return
	}
	defer clear(plaintext)
	defer reply.Clear()
	if !json.Valid(plaintext) || len(plaintext) == 0 || plaintext[0] != '{' {
		s.writeError(response, service.ErrValidation)
		return
	}
	inner := request.Request.Clone(request.Request.Context())
	inner.Body = io.NopCloser(bytes.NewReader(plaintext))
	inner.ContentLength = int64(len(plaintext))
	originalRequest := request.Request
	request.Request = inner
	defer func() { request.Request = originalRequest }()
	buffer := &transportResponseWriter{header: response.Header().Clone()}
	originalWriter := response.ResponseWriter
	response.ResponseWriter = buffer
	defer func() { response.ResponseWriter = originalWriter; clear(buffer.body.Bytes()) }()
	chain.ProcessFilter(request, response)
	if buffer.status == 0 {
		buffer.status = http.StatusOK
	}
	if buffer.err != nil {
		buffer.status = http.StatusInternalServerError
		clear(buffer.body.Bytes())
		buffer.body.Reset()
		_, _ = buffer.body.WriteString(`{"error":"response is too large"}`)
	}
	sealed, err := reply.Seal(buffer.status, buffer.body.Bytes())
	if err != nil {
		originalWriter.WriteHeader(http.StatusInternalServerError)
		return
	}
	for key, values := range buffer.header {
		originalWriter.Header()[key] = values
	}
	originalWriter.Header().Del("Content-Length")
	originalWriter.Header().Del("Content-Encoding")
	originalWriter.Header().Set("Content-Type", "application/json")
	originalWriter.Header().Set("X-AG-Encrypted", "1")
	originalWriter.WriteHeader(buffer.status)
	_ = json.NewEncoder(originalWriter).Encode(sealed)
}

type transportResponseWriter struct {
	header http.Header
	body   bytes.Buffer
	status int
	err    error
}

func (w *transportResponseWriter) Header() http.Header { return w.header }
func (w *transportResponseWriter) WriteHeader(status int) {
	if w.status == 0 {
		w.status = status
	}
}
func (w *transportResponseWriter) Write(data []byte) (int, error) {
	if w.body.Len()+len(data) > 32<<20 {
		w.err = errors.New("encrypted response exceeds limit")
		return 0, w.err
	}
	w.WriteHeader(http.StatusOK)
	return w.body.Write(data)
}
